File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/services/Network/HTTPService.ts
/**
 * HTTPService.ts
 *
 * Core HTTP service implementation using fetch (built-in to React Native).
 * All network logic is centralized here.
 *
 * This is analogous to Swift's URLSession - using the platform's native HTTP client.
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Data/Network/Services/HTTPService.swift
 */

// React Native global types
declare const fetch: (url: string, init?: RequestInit) => Promise<Response>;
declare const setTimeout: (callback: () => void, ms: number) => number;
declare const clearTimeout: (id: number) => void;
declare const AbortController: {
  new (): {
    signal: AbortSignal;
    abort(): void;
  };
};

interface RequestInit {
  method?: string;
  headers?: Record<string, string>;
  body?: string;
  signal?: AbortSignal;
}

interface AbortSignal {
  aborted: boolean;
}

interface Response {
  ok: boolean;
  status: number;
  statusText: string;
  text(): Promise<string>;
  json(): Promise<unknown>;
}

import { SDKLogger } from '../../Foundation/Logging/Logger/SDKLogger';
import { SDKError } from '../../Foundation/ErrorTypes';
import { ErrorCode } from '../../Foundation/ErrorTypes/ErrorCodes';
import { SDKConstants } from '../../Foundation/Constants';

const logger = new SDKLogger('HTTPService');

// SDK Constants - use centralized constants where available
const SDK_CLIENT = 'RunAnywhereSDK';
const SDK_PLATFORM = 'react-native';
const DEFAULT_TIMEOUT_MS = 30000;

/**
 * SDK Environment enum matching Swift/C++ SDKEnvironment
 * Uses string values to match types/enums.ts
 */
export enum SDKEnvironment {
  Development = 'development',
  Staging = 'staging',
  Production = 'production',
}

/**
 * HTTP Service Configuration
 */
export interface HTTPServiceConfig {
  /** Base URL for API requests */
  baseURL: string;
  /** API key for authentication */
  apiKey: string;
  /** SDK environment */
  environment: SDKEnvironment;
  /** Request timeout in milliseconds */
  timeoutMs?: number;
}

/**
 * Development (Supabase) Configuration
 */
export interface DevModeConfig {
  /** Supabase project URL */
  supabaseURL: string;
  /** Supabase anon key */
  supabaseKey: string;
}

/**
 * HTTP Service - Core network implementation using fetch
 *
 * Centralized HTTP transport layer modeled after Swift's URLSession approach.
 * Uses fetch - the built-in HTTP client in React Native.
 *
 * Features:
 * - Environment-aware routing (Supabase for dev, Railway for prod)
 * - Automatic header management
 * - Proper timeout and error handling
 * - Device registration with Supabase UPSERT support
 *
 * Usage:
 * ```typescript
 * // Configure (called during SDK init)
 * HTTPService.shared.configure({
 *   baseURL: 'https://api.runanywhere.ai',
 *   apiKey: 'your-api-key',
 *   environment: SDKEnvironment.Production,
 * });
 *
 * // Make requests
 * const response = await HTTPService.shared.post('/api/v1/devices/register', deviceData);
 * ```
 */
export class HTTPService {
  // ============================================================================
  // Singleton
  // ============================================================================

  private static _instance: HTTPService | null = null;

  /**
   * Get shared HTTPService instance
   */
  static get shared(): HTTPService {
    if (!HTTPService._instance) {
      HTTPService._instance = new HTTPService();
    }
    return HTTPService._instance;
  }

  // ============================================================================
  // Configuration
  // ============================================================================

  private baseURL: string = '';
  private apiKey: string = '';
  private environment: SDKEnvironment = SDKEnvironment.Production;
  private accessToken: string | null = null;
  private timeoutMs: number = DEFAULT_TIMEOUT_MS;

  // Development mode (Supabase)
  private supabaseURL: string = '';
  private supabaseKey: string = '';

  // ============================================================================
  // Initialization
  // ============================================================================

  private constructor() {}

  private get defaultHeaders(): Record<string, string> {
    return {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      'X-SDK-Client': SDK_CLIENT,
      'X-SDK-Version': SDKConstants.version,
      'X-Platform': SDK_PLATFORM,
    };
  }

  // ============================================================================
  // Configuration Methods
  // ============================================================================

  /**
   * Configure HTTP service with base URL and API key
   */
  configure(config: HTTPServiceConfig): void {
    this.baseURL = config.baseURL;
    this.apiKey = config.apiKey;
    this.environment = config.environment;
    this.timeoutMs = config.timeoutMs || DEFAULT_TIMEOUT_MS;

    logger.info(
      `Configured for ${this.getEnvironmentName()} environment: ${this.getHostname(config.baseURL)}`
    );
  }

  /**
   * Configure development mode with Supabase credentials
   *
   * When in development mode, SDK makes calls directly to Supabase
   * instead of going through the Railway backend.
   */
  configureDev(config: DevModeConfig): void {
    this.supabaseURL = config.supabaseURL;
    this.supabaseKey = config.supabaseKey;

    logger.info('Development mode configured with Supabase');
  }

  /**
   * Set authorization token
   */
  setToken(token: string): void {
    this.accessToken = token;
    logger.debug('Access token set');
  }

  /**
   * Clear authorization token
   */
  clearToken(): void {
    this.accessToken = null;
    logger.debug('Access token cleared');
  }

  /**
   * Check if HTTP service is configured
   */
  get isConfigured(): boolean {
    if (this.environment === SDKEnvironment.Development) {
      return !!this.supabaseURL;
    }
    return !!this.baseURL && !!this.apiKey;
  }

  /**
   * Get current base URL
   */
  get currentBaseURL(): string {
    if (this.environment === SDKEnvironment.Development && this.supabaseURL) {
      return this.supabaseURL;
    }
    return this.baseURL;
  }

  // ============================================================================
  // HTTP Methods
  // ============================================================================

  /**
   * POST request with JSON body
   *
   * @param path API endpoint path
   * @param data Request body (will be JSON serialized)
   * @returns Response data
   */
  async post<T = unknown, R = unknown>(path: string, data?: T): Promise<R> {
    let url = this.buildFullURL(path);

    // Handle device registration - add UPSERT for Supabase
    const isDeviceReg = this.isDeviceRegistrationPath(path);
    const headers = this.buildHeaders(isDeviceReg);

    if (isDeviceReg && this.environment === SDKEnvironment.Development) {
      const separator = url.includes('?') ? '&' : '?';
      url = `${url}${separator}on_conflict=device_id`;
    }

    const response = await this.executeRequest('POST', url, headers, data);

    // Handle 409 as success for device registration (device already exists)
    if (isDeviceReg && response.status === 409) {
      logger.info('Device already registered (409) - treating as success');
      return this.parseResponse<R>(response);
    }

    return this.handleResponse<R>(response, path);
  }

  /**
   * GET request
   *
   * @param path API endpoint path
   * @returns Response data
   */
  async get<R = unknown>(path: string): Promise<R> {
    const url = this.buildFullURL(path);
    const headers = this.buildHeaders(false);

    const response = await this.executeRequest('GET', url, headers);
    return this.handleResponse<R>(response, path);
  }

  /**
   * PUT request
   *
   * @param path API endpoint path
   * @param data Request body
   * @returns Response data
   */
  async put<T = unknown, R = unknown>(path: string, data?: T): Promise<R> {
    const url = this.buildFullURL(path);
    const headers = this.buildHeaders(false);

    const response = await this.executeRequest('PUT', url, headers, data);
    return this.handleResponse<R>(response, path);
  }

  /**
   * DELETE request
   *
   * @param path API endpoint path
   * @returns Response data
   */
  async delete<R = unknown>(path: string): Promise<R> {
    const url = this.buildFullURL(path);
    const headers = this.buildHeaders(false);

    const response = await this.executeRequest('DELETE', url, headers);
    return this.handleResponse<R>(response, path);
  }

  /**
   * POST request with raw response (returns raw data)
   *
   * @param path API endpoint path
   * @param data Request body
   * @returns Raw response data as string
   */
  async postRaw(path: string, data?: unknown): Promise<string> {
    const response = await this.post<unknown, unknown>(path, data);
    return typeof response === 'string' ? response : JSON.stringify(response);
  }

  /**
   * GET request with raw response
   *
   * @param path API endpoint path
   * @returns Raw response data as string
   */
  async getRaw(path: string): Promise<string> {
    const response = await this.get<unknown>(path);
    return typeof response === 'string' ? response : JSON.stringify(response);
  }

  // ============================================================================
  // Private Implementation
  // ============================================================================

  private async executeRequest<T>(
    method: string,
    url: string,
    headers: Record<string, string>,
    data?: T
  ): Promise<Response> {
    logger.debug(`${method} ${url}`);

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);

    try {
      const options: RequestInit = {
        method,
        headers,
        signal: controller.signal,
      };

      if (data !== undefined && method !== 'GET') {
        options.body = JSON.stringify(data);
      }

      const response = await fetch(url, options);
      return response;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private buildHeaders(isDeviceRegistration: boolean): Record<string, string> {
    const headers: Record<string, string> = { ...this.defaultHeaders };

    if (this.environment === SDKEnvironment.Development) {
      // Development mode - use Supabase headers
      // Supabase requires BOTH apikey AND Authorization: Bearer headers
      if (this.supabaseKey) {
        headers['apikey'] = this.supabaseKey;
        headers['Authorization'] = `Bearer ${this.supabaseKey}`;
        headers['Prefer'] = isDeviceRegistration
          ? 'resolution=merge-duplicates'
          : 'return=representation';
      }
    } else {
      // Production/Staging - use Bearer token
      const token = this.accessToken || this.apiKey;
      if (token) {
        headers['Authorization'] = `Bearer ${token}`;
      }
    }

    return headers;
  }

  private buildFullURL(path: string): string {
    // Handle full URLs
    if (path.startsWith('http://') || path.startsWith('https://')) {
      return path;
    }

    const base = this.currentBaseURL.replace(/\/$/, '');
    const endpoint = path.startsWith('/') ? path : `/${path}`;
    return `${base}${endpoint}`;
  }

  private isDeviceRegistrationPath(path: string): boolean {
    return (
      path.includes('sdk_devices') ||
      path.includes('devices/register') ||
      path.includes('rest/v1/sdk_devices')
    );
  }

  private async parseResponse<R>(response: Response): Promise<R> {
    const text = await response.text();
    if (!text) {
      return {} as R;
    }
    try {
      return JSON.parse(text) as R;
    } catch {
      return text as unknown as R;
    }
  }

  private async handleResponse<R>(response: Response, path: string): Promise<R> {
    if (response.ok) {
      return this.parseResponse<R>(response);
    }

    // Parse error response
    let errorMessage = `HTTP ${response.status}`;
    try {
      const errorData = (await response.json()) as Record<string, unknown>;
      errorMessage =
        (errorData.message as string) ||
        (errorData.error as string) ||
        (errorData.hint as string) ||
        errorMessage;
    } catch {
      // Ignore JSON parse errors
    }

    logger.error(`HTTP ${response.status}: ${path}`);
    throw this.createError(response.status, errorMessage, path);
  }

  private createError(statusCode: number, message: string, path: string): SDKError {
    switch (statusCode) {
      case 400:
        return new SDKError(ErrorCode.InvalidInput, `Bad request: ${message}`);
      case 401:
        return new SDKError(ErrorCode.AuthenticationFailed, message);
      case 403:
        return new SDKError(ErrorCode.AuthenticationFailed, `Forbidden: ${message}`);
      case 404:
        return new SDKError(ErrorCode.ApiError, `Not found: ${path}`);
      case 429:
        return new SDKError(ErrorCode.NetworkTimeout, `Rate limited: ${message}`);
      case 500:
      case 502:
      case 503:
      case 504:
        return new SDKError(ErrorCode.ApiError, `Server error (${statusCode}): ${message}`);
      default:
        return new SDKError(ErrorCode.NetworkUnavailable, `HTTP ${statusCode}: ${message}`);
    }
  }

  private getEnvironmentName(): string {
    switch (this.environment) {
      case SDKEnvironment.Development:
        return 'development';
      case SDKEnvironment.Staging:
        return 'staging';
      case SDKEnvironment.Production:
        return 'production';
      default:
        return 'unknown';
    }
  }

  private getHostname(url: string): string {
    // Simple hostname extraction for React Native compatibility
    const match = url.match(/^https?:\/\/([^/:]+)/);
    return match ? match[1] : url.substring(0, 30);
  }
}

export default HTTPService;

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/data/network/http_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' as http;
import 'package:runanywhere/data/network/network_configuration.dart';
import 'package:runanywhere/foundation/configuration/sdk_constants.dart';
import 'package:runanywhere/foundation/error_types/sdk_error.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/native/dart_bridge_auth.dart';
import 'package:runanywhere/public/configuration/sdk_environment.dart';

/// HTTP Service - Core network implementation using dart:http
///
/// Centralized HTTP transport layer matching Swift/React Native HTTPService.
/// Uses the http package as the HTTP client.
///
/// Features:
/// - Environment-aware routing (Supabase for dev, Railway for prod)
/// - Automatic header management
/// - Proper timeout and error handling
/// - Device registration with Supabase UPSERT support
///
/// Usage:
/// ```dart
/// // Configure (called during SDK init)
/// HTTPService.shared.configure(HTTPServiceConfig(
///   baseURL: 'https://api.runanywhere.ai',
///   apiKey: 'your-api-key',
///   environment: SDKEnvironment.production,
/// ));
///
/// // Make requests
/// final response = await HTTPService.shared.post('/api/v1/devices/register', deviceData);
/// ```
class HTTPService {
  // ============================================================================
  // Singleton
  // ============================================================================

  static HTTPService? _instance;

  /// Get shared HTTPService instance
  static HTTPService get shared {
    _instance ??= HTTPService._();
    return _instance!;
  }

  // ============================================================================
  // Configuration
  // ============================================================================

  String _baseURL = '';
  String _apiKey = '';
  SDKEnvironment _environment = SDKEnvironment.production;
  String? _accessToken;
  Duration _timeout = const Duration(seconds: 30);

  // Development mode (Supabase)
  String _supabaseURL = '';
  String _supabaseKey = '';

  final http.Client _httpClient;
  final SDKLogger _logger;

  // ============================================================================
  // Initialization
  // ============================================================================

  HTTPService._()
      : _httpClient = http.Client(),
        _logger = SDKLogger('HTTPService');

  Map<String, String> get _defaultHeaders => {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'X-SDK-Client': 'RunAnywhereFlutterSDK',
        'X-SDK-Version': SDKConstants.version,
        'X-Platform': SDKConstants.platform,
      };

  // ============================================================================
  // Configuration Methods
  // ============================================================================

  /// Configure HTTP service with base URL and API key
  void configure(HTTPServiceConfig config) {
    _baseURL = config.baseURL;
    _apiKey = config.apiKey;
    _environment = config.environment;
    _timeout = Duration(milliseconds: config.timeoutMs);

    _logger.info(
      'Configured for ${_getEnvironmentName()} environment: ${_getHostname(config.baseURL)}',
    );
  }

  /// Configure development mode with Supabase credentials
  ///
  /// When in development mode, SDK makes calls directly to Supabase
  /// instead of going through the Railway backend.
  void configureDev(DevModeConfig config) {
    _supabaseURL = config.supabaseURL;
    _supabaseKey = config.supabaseKey;

    _logger.info('Development mode configured with Supabase');
  }

  /// Set authorization token
  void setToken(String token) {
    _accessToken = token;
    _logger.debug('Access token set');
  }

  /// Clear authorization token
  void clearToken() {
    _accessToken = null;
    _logger.debug('Access token cleared');
  }

  /// Check if HTTP service is configured
  bool get isConfigured {
    if (_environment == SDKEnvironment.development) {
      return _supabaseURL.isNotEmpty;
    }
    return _baseURL.isNotEmpty && _apiKey.isNotEmpty;
  }

  // ============================================================================
  // Token Resolution (matches Swift's resolveToken)
  // ============================================================================

  /// Resolve valid token for request, refreshing if needed.
  /// Matches Swift's HTTPService.resolveToken(requiresAuth:)
  Future<String> _resolveToken({required bool requiresAuth}) async {
    if (_environment == SDKEnvironment.development) {
      // Development mode - use Supabase key directly
      return _supabaseKey;
    }

    if (!requiresAuth) {
      // Non-auth requests use API key
      return _apiKey;
    }

    // Production/Staging - check for valid token, refresh if needed
    final authBridge = DartBridgeAuth.instance;

    // Check if we have a valid token
    final currentToken = authBridge.getAccessToken();
    if (currentToken != null && !authBridge.needsRefresh()) {
      return currentToken;
    }

    // Try refresh if we have a refresh token
    if (authBridge.isAuthenticated()) {
      _logger.debug('Token needs refresh, attempting refresh...');
      final result = await authBridge.refreshToken();
      if (result.isSuccess) {
        final newToken = authBridge.getAccessToken();
        if (newToken != null) {
          // Update internal access token
          _accessToken = newToken;
          _logger.info('Token refreshed successfully');
          return newToken;
        }
      } else {
        _logger.warning('Token refresh failed: ${result.error}');
      }
    }

    // Fallback to access token or API key
    if (_accessToken != null && _accessToken!.isNotEmpty) {
      return _accessToken!;
    }
    if (_apiKey.isNotEmpty) {
      return _apiKey;
    }

    throw SDKError.authenticationFailed('No valid authentication token');
  }

  /// Get current base URL
  String get currentBaseURL {
    if (_environment == SDKEnvironment.development && _supabaseURL.isNotEmpty) {
      return _supabaseURL;
    }
    return _baseURL;
  }

  /// Get current environment
  SDKEnvironment get environment => _environment;

  // ============================================================================
  // HTTP Methods
  // ============================================================================

  /// POST request with JSON body
  ///
  /// [path] - API endpoint path
  /// [data] - Request body (will be JSON serialized)
  /// Returns parsed response data
  Future<T> post<T>(
    String path,
    Object? data, {
    T Function(Map<String, dynamic>)? fromJson,
    bool requiresAuth = false,
  }) async {
    var url = _buildFullURL(path);

    // Handle device registration - add UPSERT for Supabase
    final isDeviceReg = _isDeviceRegistrationPath(path);
    final headers = _buildHeaders(isDeviceReg, requiresAuth);

    if (isDeviceReg && _environment == SDKEnvironment.development) {
      final separator = url.contains('?') ? '&' : '?';
      url = '$url${separator}on_conflict=device_id';
    }

    final response = await _executeRequest(
      'POST',
      url,
      headers,
      data,
      requiresAuth: requiresAuth,
    );

    // Handle 409 as success for device registration (device already exists)
    if (isDeviceReg && response.statusCode == 409) {
      _logger.info('Device already registered (409) - treating as success');
      return _parseResponse<T>(response, fromJson);
    }

    return _handleResponse<T>(response, path, fromJson);
  }

  /// POST request returning raw bytes
  Future<Uint8List> postRaw(
    String path,
    Uint8List payload, {
    bool requiresAuth = false,
  }) async {
    var url = _buildFullURL(path);

    final isDeviceReg = _isDeviceRegistrationPath(path);
    final headers = _buildHeaders(isDeviceReg, requiresAuth);

    if (isDeviceReg && _environment == SDKEnvironment.development) {
      final separator = url.contains('?') ? '&' : '?';
      url = '$url${separator}on_conflict=device_id';
    }

    final uri = Uri.parse(url);
    _logger.debug('POST $path');

    try {
      final response = await _httpClient
          .post(
            uri,
            headers: headers,
            body: payload,
          )
          .timeout(_timeout);

      if (isDeviceReg && response.statusCode == 409) {
        _logger.info('Device already registered (409) - treating as success');
        return response.bodyBytes;
      }

      _validateResponse(response, path);
      return response.bodyBytes;
    } catch (e) {
      if (e is SDKError) rethrow;
      _logger.error('POST $path failed: $e');
      throw SDKError.networkError(e.toString());
    }
  }

  /// GET request
  ///
  /// [path] - API endpoint path
  /// Returns parsed response data
  Future<T> get<T>(
    String path, {
    T Function(Map<String, dynamic>)? fromJson,
    bool requiresAuth = false,
  }) async {
    final url = _buildFullURL(path);
    final headers = _buildHeaders(false, requiresAuth);

    final response = await _executeRequest(
      'GET',
      url,
      headers,
      null,
      requiresAuth: requiresAuth,
    );
    return _handleResponse<T>(response, path, fromJson);
  }

  /// GET request returning raw bytes
  Future<Uint8List> getRaw(
    String path, {
    bool requiresAuth = false,
  }) async {
    final url = _buildFullURL(path);
    final headers = _buildHeaders(false, requiresAuth);

    final uri = Uri.parse(url);
    _logger.debug('GET $path');

    try {
      final response = await _httpClient
          .get(
            uri,
            headers: headers,
          )
          .timeout(_timeout);

      _validateResponse(response, path);
      return response.bodyBytes;
    } catch (e) {
      if (e is SDKError) rethrow;
      _logger.error('GET $path failed: $e');
      throw SDKError.networkError(e.toString());
    }
  }

  /// PUT request
  ///
  /// [path] - API endpoint path
  /// [data] - Request body
  /// Returns parsed response data
  Future<T> put<T>(
    String path,
    Object? data, {
    T Function(Map<String, dynamic>)? fromJson,
    bool requiresAuth = false,
  }) async {
    final url = _buildFullURL(path);
    final headers = _buildHeaders(false, requiresAuth);

    final response = await _executeRequest(
      'PUT',
      url,
      headers,
      data,
      requiresAuth: requiresAuth,
    );
    return _handleResponse<T>(response, path, fromJson);
  }

  /// DELETE request
  ///
  /// [path] - API endpoint path
  /// Returns parsed response data
  Future<T> delete<T>(
    String path, {
    T Function(Map<String, dynamic>)? fromJson,
    bool requiresAuth = false,
  }) async {
    final url = _buildFullURL(path);
    final headers = _buildHeaders(false, requiresAuth);

    final response = await _executeRequest(
      'DELETE',
      url,
      headers,
      null,
      requiresAuth: requiresAuth,
    );
    return _handleResponse<T>(response, path, fromJson);
  }

  // ============================================================================
  // Private Implementation
  // ============================================================================

  Future<http.Response> _executeRequest(
    String method,
    String url,
    Map<String, String> headers,
    Object? data, {
    bool requiresAuth = false,
    bool isRetry = false,
  }) async {
    final uri = Uri.parse(url);
    _logger.debug('$method $url');

    try {
      // Resolve auth token if required (matches Swift's resolveToken pattern)
      if (requiresAuth && _environment != SDKEnvironment.development) {
        final token = await _resolveToken(requiresAuth: requiresAuth);
        if (token.isNotEmpty) {
          headers['Authorization'] = 'Bearer $token';
        }
      }

      late http.Response response;

      switch (method) {
        case 'GET':
          response = await _httpClient.get(uri, headers: headers).timeout(_timeout);
          break;
        case 'POST':
          final body = data != null ? json.encode(data) : null;
          // Debug: Log request body for telemetry debugging
          if (url.contains('telemetry')) {
            _logger.debug('POST body: $body');
          }
          response = await _httpClient
              .post(
                uri,
                headers: headers,
                body: body,
              )
              .timeout(_timeout);
          // Debug: Log response for telemetry debugging
          if (url.contains('telemetry')) {
            _logger.debug('Response status: ${response.statusCode}');
            _logger.debug('Response body: ${response.body}');
          }
          break;
        case 'PUT':
          response = await _httpClient
              .put(
                uri,
                headers: headers,
                body: data != null ? json.encode(data) : null,
              )
              .timeout(_timeout);
          break;
        case 'DELETE':
          response = await _httpClient.delete(uri, headers: headers).timeout(_timeout);
          break;
        default:
          throw SDKError.networkError('Unsupported HTTP method: $method');
      }

      // Handle 401 Unauthorized - attempt token refresh and retry once
      if (response.statusCode == 401 && requiresAuth && !isRetry) {
        _logger.debug('Received 401, attempting token refresh and retry...');
        
        final authBridge = DartBridgeAuth.instance;
        final refreshResult = await authBridge.refreshToken();
        
        if (refreshResult.isSuccess) {
          final newToken = authBridge.getAccessToken();
          if (newToken != null) {
            _accessToken = newToken;
            _logger.info('Token refreshed, retrying request...');
            
            // Retry the request with new token
            final retryHeaders = Map<String, String>.from(headers);
            return _executeRequest(
              method,
              url,
              retryHeaders,
              data,
              requiresAuth: requiresAuth,
              isRetry: true,
            );
          }
        } else {
          _logger.warning('Token refresh failed: ${refreshResult.error}');
        }
      }

      return response;
    } on TimeoutException {
      _logger.error('$method $url timed out');
      throw SDKError.timeout('Request timed out');
    } catch (e) {
      if (e is SDKError) rethrow;
      _logger.error('$method $url failed: $e');
      throw SDKError.networkError(e.toString());
    }
  }

  Map<String, String> _buildHeaders(bool isDeviceRegistration, bool requiresAuth) {
    final headers = Map<String, String>.from(_defaultHeaders);

    if (_environment == SDKEnvironment.development) {
      // Development mode - use Supabase headers
      // Supabase requires BOTH apikey AND Authorization: Bearer headers
      if (_supabaseKey.isNotEmpty) {
        headers['apikey'] = _supabaseKey;
        headers['Authorization'] = 'Bearer $_supabaseKey';
        headers['Prefer'] = isDeviceRegistration
            ? 'resolution=merge-duplicates'
            : 'return=representation';
      }
    } else {
      // Production/Staging - use Bearer token
      final token = _accessToken ?? _apiKey;
      if (token.isNotEmpty) {
        headers['Authorization'] = 'Bearer $token';
      }
      // Also add apikey for production (Railway backend)
      if (_apiKey.isNotEmpty) {
        headers['apikey'] = _apiKey;
      }
    }

    return headers;
  }

  String _buildFullURL(String path) {
    // Handle full URLs
    if (path.startsWith('http://') || path.startsWith('https://')) {
      return path;
    }

    final base = currentBaseURL.endsWith('/')
        ? currentBaseURL.substring(0, currentBaseURL.length - 1)
        : currentBaseURL;
    final endpoint = path.startsWith('/') ? path : '/$path';
    return '$base$endpoint';
  }

  bool _isDeviceRegistrationPath(String path) {
    return path.contains('sdk_devices') ||
        path.contains('devices/register') ||
        path.contains('rest/v1/sdk_devices');
  }

  T _parseResponse<T>(
    http.Response response,
    T Function(Map<String, dynamic>)? fromJson,
  ) {
    final text = response.body;
    if (text.isEmpty) {
      return {} as T;
    }
    try {
      final decoded = json.decode(text);
      if (fromJson != null && decoded is Map<String, dynamic>) {
        return fromJson(decoded);
      }
      return decoded as T;
    } catch (_) {
      return text as T;
    }
  }

  T _handleResponse<T>(
    http.Response response,
    String path,
    T Function(Map<String, dynamic>)? fromJson,
  ) {
    if (response.statusCode >= 200 && response.statusCode < 300) {
      return _parseResponse<T>(response, fromJson);
    }

    // Parse error response
    var errorMessage = 'HTTP ${response.statusCode}';
    try {
      final errorData = json.decode(response.body) as Map<String, dynamic>;
      errorMessage = (errorData['message'] as String?) ??
          (errorData['error'] as String?) ??
          (errorData['hint'] as String?) ??
          errorMessage;
    } catch (_) {
      // Ignore JSON parse errors
    }

    _logger.error('HTTP ${response.statusCode}: $path');
    throw _createError(response.statusCode, errorMessage, path);
  }

  void _validateResponse(http.Response response, String path) {
    if (response.statusCode >= 200 && response.statusCode < 300) {
      return;
    }

    var errorMessage = 'HTTP ${response.statusCode}';
    try {
      final errorData = json.decode(response.body) as Map<String, dynamic>;
      errorMessage = (errorData['message'] as String?) ??
          (errorData['error'] as String?) ??
          (errorData['hint'] as String?) ??
          errorMessage;
    } catch (_) {
      // Keep default error message if parsing fails
    }

    _logger.error('HTTP ${response.statusCode}: $path - $errorMessage');
    throw _createError(response.statusCode, errorMessage, path);
  }

  SDKError _createError(int statusCode, String message, String path) {
    switch (statusCode) {
      case 400:
        return SDKError.networkError('Bad request: $message');
      case 401:
        return SDKError.authenticationFailed(message);
      case 403:
        return SDKError.authenticationFailed('Forbidden: $message');
      case 404:
        return SDKError.networkError('Not found: $path');
      case 429:
        return SDKError.rateLimitExceeded('Rate limited: $message');
      case 500:
      case 502:
      case 503:
      case 504:
        return SDKError.serverError('Server error ($statusCode): $message');
      default:
        return SDKError.networkError('HTTP $statusCode: $message');
    }
  }

  String _getEnvironmentName() {
    switch (_environment) {
      case SDKEnvironment.development:
        return 'development';
      case SDKEnvironment.staging:
        return 'staging';
      case SDKEnvironment.production:
        return 'production';
    }
  }

  String _getHostname(String url) {
    // Simple hostname extraction
    final match = RegExp(r'^https?://([^/:]+)').firstMatch(url);
    return match != null ? match.group(1)! : url.substring(0, 30.clamp(0, url.length));
  }

  /// Reset for testing
  static void resetForTesting() {
    _instance = null;
  }
}
